examples/shell/shell_common/cmd_ping.cpp

```diff
@@ -129,12 +129,60 @@ class PingArguments
 } gPingArguments;
 
 Protocols::Echo::EchoClient gEchoClient;
+Transport::AdminPairingTable gAdmins;
 
-bool EchoIntervalExpired(void)
+CHIP_ERROR SendEchoRequest(streamer_t * stream);
+
+Transport::PeerAddress GetEchoPeerAddress()
 {
-    uint64_t now = System::Timer::GetCurrentEpoch();
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    if (gPingArguments.IsUsingTCP())
+    {
+        return Transport::PeerAddress::TCP(gDestAddr, gPingArguments.GetEchoPort());
+    }
+    else
+#endif
+    {
 
-    return (now >= gPingArguments.GetLastEchoTime() + gPingArguments.GetEchoInterval());
+        return Transport::PeerAddress::UDP(gDestAddr, gPingArguments.GetEchoPort(), INET_NULL_INTERFACEID);
+    }
+}
+
+void Shutdown()
+{
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    if (gPingArguments.IsUsingTCP())
+    {
+        gTCPManager.Disconnect(GetEchoPeerAddress());
+    }
+    gTCPManager.Close();
+#endif
+    gUDPManager.Close();
+
+    gEchoClient.Shutdown();
+    gExchangeManager.Shutdown();
+    gSessionManager.Shutdown();
+}
+
+void EchoTimerHandler(chip::System::Layer * systemLayer, void * appState, chip::System::Error error)
+{
+    if (gPingArguments.GetEchoRespCount() != gPingArguments.GetEchoCount())
+    {
+        streamer_printf(streamer_get(), "No response received\n");
+        gPingArguments.SetEchoRespCount(gPingArguments.GetEchoCount());
```

Comment:
Don't understand this logic, why we need to set resp count to request count in case of timeout? Shutdown is also not called if we fall under this case